File: 338 counting-bits/js/main.v2.js
/**
 * @param {number} num
 * @return {number[]}
 */

// 由基本算法优化后的算法，
// Leetcode最优解

var countBits = function(num) {
  if (num < 0) return []

  const res = Array(num + 1).fill(0)
  for (let i = 1; i <= num; i++) {
    res[i] = res[i >> 1] + (i & 1)
  }

  return res
};

File: 374 guess-number-higher-or-lower/c/main.v2.c
// Forward declaration of guess API.
// @param num, your guess
// @return -1 if my number is lower, 1 if my number is higher, otherwise return 0

int guess(int num);

/**
 * Leetcode 最优解
 */
int guessNumber(int n)
{
  int lo = 1;
  int hi = n;
  while (lo < hi) {
    int mid = lo + (hi - lo) / 2;
    int g = guess(mid);
    if (g == 0)
      return mid;
    if (g == -1)
      hi = mid - 1;
    if (g == 1)
      lo = mid + 1;
  }
  return lo;
}
